Source/Core/DolphinQt2/Resources.cpp

```diff
@@ -2,46 +2,97 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include <algorithm>
+
+#include <QGuiApplication>
+#include <QIcon>
+#include <QPixmap>
+#include <QScreen>
 #include <QStringList>
 
 #include "Common/CommonPaths.h"
 #include "Common/FileUtil.h"
 #include "DolphinQt2/Resources.h"
+#include "DolphinQt2/Settings.h"
 
 QList<QPixmap> Resources::m_platforms;
 QList<QPixmap> Resources::m_countries;
 QList<QPixmap> Resources::m_ratings;
 QList<QPixmap> Resources::m_misc;
 
+QIcon Resources::GetIcon(const QString& name, const QString& dir)
+{
+  QString base_path = dir + name;
+
+  const auto dpr = QGuiApplication::primaryScreen()->devicePixelRatio();
+
+  QIcon icon(base_path.append(QStringLiteral(".png")));
+
+  if (dpr > 2)
+  {
+    QPixmap pixmap(base_path.append(QStringLiteral("@4x.png")));
+    if (!pixmap.isNull())
+    {
+      pixmap.setDevicePixelRatio(4.0);
+      icon.addPixmap(pixmap);
+    }
+  }
+
+  return icon;
+}
+
+QPixmap Resources::GetPixmap(const QString& name, const QString& dir)
+{
+  const auto icon = GetIcon(name, dir);
+  const auto sizes = icon.availableSizes();
+  const auto dpr = QGuiApplication::primaryScreen()->devicePixelRatio();
+
+  return icon.pixmap(sizes[std::min((dpr <= 1) ? 1 : (dpr <= 2) ? 2 : 4, sizes.size() - 1)]);
```

Comment:
This breaks on macOS (icons are too large). I think the right thing to do here is `icon.pixmap(sizes[0])`?